Assets/MixedRealityToolkit.SDK/Experimental/Features/Dwell/DwellHandler.cs

```diff
@@ -0,0 +1,210 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+using Microsoft.MixedReality.Toolkit.Input;
+using System;
+using UnityEngine;
+using UnityEngine.Events;
+
+namespace Microsoft.MixedReality.Toolkit.Experimental.Dwell
+{
+    [Serializable]
+    public class PointerUnityEvent : UnityEvent<IMixedRealityPointer>
+    {
+    }
+
+    /// <summary>
+    /// Use this component to add a Dwell modality (https://docs.microsoft.com/en-us/windows/mixed-reality/gaze-and-dwell) to the UI target.
+    /// </summary>
+    public class DwellHandler : MonoBehaviour, IMixedRealityFocusHandler
+    {
+        protected enum DwellState
+        {
+            None = 0,
+            FocusGained,
+            DwellIntended,
+            DwellStarted,
+            DwellCompleted,
+            DwellCanceled,
+            Invalid
+        }
+
+        [Header("Dwell Settings")]
+        [SerializeField]
+        protected DwellProfile dwellProfile = null;
+
+        [Header("Dwell Events")]
+        [SerializeField]
+        private PointerUnityEvent DwellIntended = new PointerUnityEvent();
+
+        [SerializeField]
+        private PointerUnityEvent DwellStarted = new PointerUnityEvent();
+
+        [SerializeField]
+        private PointerUnityEvent DwellCompleted = new PointerUnityEvent();
+
+        [SerializeField]
+        private PointerUnityEvent DwellCanceled = new PointerUnityEvent();
+
+        /// <summary>
+        /// Invoke this event if the dwell is interrupted with alternate modality
+        /// </summary>
+        public UnityAction CancelDwellAction;
+
+        protected bool HasFocus { get; private set; }
+
+        /// <summary>
+        /// Cached pointer reference to track focus events maps to teh same pointer that initiated dwell
+        /// </summary>
+        private IMixedRealityPointer pointerRef;
+
+        private DateTime focusEnterTime = DateTime.MaxValue;
+        private DateTime focusExitTime = DateTime.MaxValue;
+
+        protected DwellState currentDwellState = DwellState.None;
+        protected float fillTimer = 0;
+
+        private void Awake()
+        {
+            Debug.Assert(dwellProfile != null, "DwellProfile is null, creating default profile.");
+
+            if (dwellProfile == null)
+            {
+                dwellProfile = ScriptableObject.CreateInstance("DwellProfile") as DwellProfile;
```

Comment:
> ```cs
> dwellProfile = ScriptableObject.CreateInstance<DwellProfile>();
> ```

That will create compile issue since the  ScriptableObject.CreateInstance returns a scriptable object.

Reply:
DwellProfile is a ScriptableObject, and CreateInstance<T> specifically returns a T. If there is a problem, I'm not understanding what it is.

Reply:
Oh wait, it looks like despite the above comment you did actually change this. Re-resolving. 🙂